document_wopi/models/document_wopi_access_token.py

```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+# © 2017 Therp BV <http://therp.nl>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+import time
+import uuid
+import urllib
+import urlparse
+from datetime import timedelta
+from openerp import api, fields, models
+from openerp.exceptions import AccessDenied
+
+
+class DocumentWopiAccessToken(models.Model):
+    _name = 'document.wopi.access.token'
+    _description = 'Access token for WOPI requests'
+
+    user_id = fields.Many2one(
+        'res.users', 'User', required=True, ondelete='cascade',
+    )
+    attachment_id = fields.Many2one(
+        'ir.attachment', 'Attachment', required=True, ondelete='cascade',
+    )
+    valid_until = fields.Datetime()
+    token = fields.Char(required=True)
+    token_ttl = fields.Integer(compute='_compute_token_ttl')
+    action_url = fields.Char(compute='_compute_action_url')
+
+    @api.model
+    def _get_token(self, attachment):
+        """Create or return a token for some attachment for the current user"""
+        existing = self.search([
+            ('user_id', '=', self.env.user.id),
+            ('attachment_id', '=', attachment.id),
+            '|',
+            ('valid_until', '=', False),
+            ('valid_until', '>=', fields.Datetime.now()),
+        ], limit=1)
+        if existing:
+            return existing
+        return self.create({
+            'user_id': self.env.user.id,
+            'attachment_id': attachment.id,
+            'valid_until': fields.Datetime.to_string(
+                fields.Datetime.from_string(fields.Datetime.now()) +
+                timedelta(self.env['ir.config_parameter'].get_param(
+                    'document_wopi.token_validity', 36000,
+                ))
+            ),
+            # TODO: is this random enough or do we need some secret generator?
```

Comment:
This is the way Odoo SA does it, but I don't think it's random enough personally. 

Probably best to not diverge though, so we can likely remove this comment? 

Fail2Ban4Lyfe, although this does remind me to put up my RFC for an internal 404 trap too.